internal/databases/mysql/mysql_binlog.go

```diff
@@ -160,3 +166,151 @@ func BinlogPrefix(filename string) string {
+		Flavor:         flavor,
+		RawModeEnabled: true,
+	}
+
+	// if you specify a CA cert in the wal-g settings, it should act like we
+	// specified "&tls=custom"
+	caFile, ok := internal.GetSetting(internal.MysqlSslCaSetting)
+	if ok {
+		driverConfig.TLSConfig = "custom"
+	}
+
+	switch driverConfig.TLSConfig {
+	case "true":
+		cfg.TLSConfig = &tls.Config{}
+	case "skip-verify":
+		cfg.TLSConfig = &tls.Config{InsecureSkipVerify: true}
+	case "custom":
+		rootCertPool := x509.NewCertPool()
+		pem, err := os.ReadFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		if ok := rootCertPool.AppendCertsFromPEM(pem); !ok {
+			return nil, fmt.Errorf("failed to load certificate from %s", caFile)
+		}
+		cfg.TLSConfig = &tls.Config{RootCAs: rootCertPool}
+	}
+
+	return cfg, err
+}
+
+// NewBinlogStream begins retrieval of a raw binary log from the MySQL server
+func NewBinlogStream(db *sql.DB, binlog string) (*BinlogStream, error) {
+	flavor, err := getMySQLFlavor(db)
+	if err != nil {
+		return nil, err
+	}
+
+	datasourceName, err := internal.GetRequiredSetting(internal.MysqlDatasourceNameSetting)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg, err := GetBinlogSyncerConfig(datasourceName, flavor)
+	if err != nil {
+		return nil, err
+	}
+
+	// the go-mysql integrated logging library is super noisy
+	log.SetLevel(log.LevelError)
+	stream := BinlogStream{BinlogSyncer: *replication.NewBinlogSyncer(*cfg)}
+
+	// even if server is using GTIDs we want a specific file at offset 0
+	// (we're trying to sync this specific binlog file)
+	streamer, err := stream.StartSync(mysql.Position{Name: binlog, Pos: 0})
+	if err != nil {
+		return nil, err
+	}
+	stream.BinlogStreamer = *streamer
+	return &stream, nil
+}
+
+// Read converts the per-event output from BinlogStreamer.GetEvent() to the []byte
+// output expected by the io.Reader interface.
+func (b *BinlogStream) Read(p []byte) (int, error) {
+	if b.lastData == nil {
+		// no existing data left from a prior binlog event, read next event
+		ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+		event, err := b.GetEvent(ctx)
+		cancel()
+		if err != nil {
+			return 0, err
+		}
+
+		// handle special event types
+		switch event.Header.EventType {
+		case replication.ROTATE_EVENT:
+			if event.Header.Timestamp == 0 || event.Header.LogPos == 0 {
+				// fake rotate event, ignore
+				return 0, nil
+			}
+		case replication.HEARTBEAT_EVENT:
+			// heartbeat events aren't "real data" according to
+			// https://dev.mysql.com/doc/internals/en/heartbeat-event.html
+			// and break mysqlbinlog, ignore
+			return 0, nil
+		case replication.FORMAT_DESCRIPTION_EVENT:
+			// beginning of binlog
+			if b.hasStarted {
+				// we are now reading binlog #2, stop.
+				return 0, io.EOF
+			}
+			// start the stream by writing the binlog header
+			b.hasStarted = true
+			b.lastData = event.RawData
+			return copy(p, replication.BinLogFileHeader), nil
+		}
+
+		n := copy(p, event.RawData)
```

Comment:
Does `event.RawData` contain all the event data ?

Dummy queries like  `delete from huge_table` or `update huge_table set a = a+1` may create huge events of gigabytes in size.
Parsing such events will lead to Out-Of-Memory errros.